runtime/compiler/runtime/JITServerAOTCache.cpp

```diff
@@ -745,6 +1114,338 @@ JITServerAOTCache::printStats(FILE *f) const
+      return false;
+
+   if (!writeAOTRecordList(f, _classLoaderHead, header._numClassLoaderRecords))
+      return false;
+   if (!writeAOTRecordList(f, _classHead, header._numClassRecords))
+      return false;
+   if (!writeAOTRecordList(f, _methodHead, header._numMethodRecords))
+      return false;
+   if (!writeAOTRecordList(f, _classChainHead, header._numClassChainRecords))
+      return false;
+   if (!writeAOTRecordList(f, _wellKnownClassesHead, header._numWellKnownClassesRecords))
+      return false;
+   if (!writeAOTRecordList(f, _aotHeaderHead, header._numAOTHeaderRecords))
+      return false;
+   if (!writeCachedMethodList(f, _cachedMethodHead, header._numCachedAOTMethods))
+      return false;
+
+   return true;
+   }
+
+// Tests whether or not the given AOT snapshot is compatible with the server.
+static bool
+isCompatibleSnapshotVersion(const JITServerAOTCacheVersion &version)
+   {
+   JITServerAOTCacheVersion currentVersion = {0};
+   getCurrentAOTCacheVersion(currentVersion);
+
+   return (0 == memcmp(version._eyeCatcher, currentVersion._eyeCatcher, sizeof(currentVersion._eyeCatcher))) &&
+          (version._snapshotVersion == currentVersion._snapshotVersion) &&
+          (version._jitserverVersion == currentVersion._jitserverVersion);
+   }
+
+// Read an AOT cache snapshot, returning NULL if the cache is ill-formed or
+// incompatible with the running server.
+JITServerAOTCache *
+JITServerAOTCache::readCache(FILE *f, const std::string &name, TR_Memory &trMemory)
+   {
+   if (!JITServerAOTCacheMap::cacheHasSpace())
+      return NULL;
+
+   JITServerAOTCacheHeader header = {0};
+   if (1 != fread(&header, sizeof(JITServerAOTCacheHeader), 1, f))
+      return NULL;
+
+   if (!isCompatibleSnapshotVersion(header._version))
+      return NULL;
+
+   JITServerAOTCache *cache = new (TR::Compiler->persistentGlobalMemory()) JITServerAOTCache(name);
+   if (!cache)
+      return NULL;
+
+   if (!cache->readCache(f, header, trMemory))
+      {
+      cache->~JITServerAOTCache();
+      TR::Compiler->persistentGlobalMemory()->freePersistentMemory(cache);
+      return NULL;
+      }
+
+   return cache;
+   }
+
+bool
+JITServerAOTCache::readCache(FILE *f, const JITServerAOTCacheHeader &header, TR_Memory &trMemory)
+   {
+   _classLoaderMap.reserve(header._numClassLoaderRecords);
+   _classMap.reserve(header._numClassRecords);
+   _methodMap.reserve(header._numMethodRecords);
+   _classChainMap.reserve(header._numClassChainRecords);
+   _wellKnownClassesMap.reserve(header._numWellKnownClassesRecords);
+   _aotHeaderMap.reserve(header._numAOTHeaderRecords);
+   _cachedMethodMap.reserve(header._numCachedAOTMethods);
+
+   _nextClassLoaderId = header._nextClassLoaderId;
+   _nextClassId = header._nextClassId;
+   _nextMethodId = header._nextMethodId;
+   _nextClassChainId = header._nextClassChainId;
+   _nextWellKnownClassesId = header._nextWellKnownClassesId;
+   _nextAOTHeaderId = header._nextAOTHeaderId;
+
```

Comment:
Should instantiate a local `TR::StackMemoryRegion` object here and pass it to allocator constructors.